tests/test_dimensions.py

```diff
@@ -212,5 +251,388 @@ def testPickling(self):
             self.assertIs(graph1, graph2)
 
 
+@dataclass
+class SplitByStateFlags:
+    """A struct that separates data IDs with different states but the same
+    values.
+    """
+
+    minimal: Optional[DataCoordinateSequence] = None
+    """Data IDs that only contain values for required dimensions.
+
+    `DataCoordinateSequence.hasFull()` will return `True` for this if and only
+    if ``minimal.graph.implied`` has no elements.
+    `DataCoordinate.hasRecords()` will always return `False`.
+    """
+
+    complete: Optional[DataCoordinateSequence] = None
+    """Data IDs that contain values for all dimensions.
+
+    `DataCoordinateSequence.hasFull()` will always `True` and
+    `DataCoordinate.hasRecords()` will always return `True` for this attribute.
+    """
+
+    expanded: Optional[DataCoordinateSequence] = None
+    """Data IDs that contain values for all dimensions as well as records.
+
+    `DataCoordinateSequence.hasFull()` and `DataCoordinate.hasRecords()` will
+    always return `True` for this attribute.
+    """
+
+    def chain(self, n: Optional[int] = None) -> Iterator:
+        """Iterate over the data IDs of different types.
+
+        Parameters
+        ----------
+        n : `int`, optional
+            If provided (`None` is default), iterate over only the ``nth``
+            data ID in each attribute.
+
+        Yields
+        ------
+        dataId : `DataCoordinate`
+            A data ID from one of the attributes in this struct.
+        """
+        if n is None:
+            s = slice(None, None)
+        else:
+            s = slice(n, n + 1)
+        if self.minimal is not None:
+            yield from self.minimal[s]
+        if self.complete is not None:
+            yield from self.complete[s]
+        if self.expanded is not None:
+            yield from self.expanded[s]
+
+
+class DataCoordinateTestCase(unittest.TestCase):
+
+    RANDOM_SEED = 10
+
+    @classmethod
+    def setUpClass(cls):
+        cls.allDataIds = loadDimensionData()
+
+    def setUp(self):
+        self.rng = Random(self.RANDOM_SEED)
+
+    def randomDataIds(self, n: int, dataIds: Optional[DataCoordinateSequence] = None):
+        """Select random data IDs from those loaded from test data.
+
+        Parameters
+        ----------
+        n : `int`
+             Number of data IDs to select.
+        dataIds : `DataCoordinateSequence`, optional
+            Data IDs to select from.  Defaults to ``self.allDataIds``.
+
+        Returns
+        -------
+        selected : `DataCoordinateSequence`
+            ``n`` Data IDs randomly selected from ``dataIds`` with replacement.
+        """
+        if dataIds is None:
+            dataIds = self.allDataIds
+        return DataCoordinateSequence(self.rng.sample(dataIds, n),
```

Comment:
This reminded me that we have code in `python/lsst/daf/butler/tests/_testRepo.py` that uses numpy for this for no reason at all (it's the only explicit usage of numpy in the butler) where `rng.sample(range(10), 10)` would work fine.